src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,214 @@
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(T value) {
+            this.value = value;
+            this.next = null;
+            this.prev = null;
+        }
+    }
+
+    void checkIndexValid(int index) {
+        if (index < 0) {
+            throw new IndexOutOfBoundsException("Index must be greater then 0 but was :" + index);
+        }
+
+        if (index > size - 1) {
+            throw new IndexOutOfBoundsException("Index can't be higher then size but Index was :"
+                    + index + " and Size was " + size + " no such index in Node");
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+
+        if (size == 0) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + " Size: " + size);
+        }
+
+        Node<T> newNode = new Node<>(value);
+
+        if (index == 0) {
+            newNode.next = head;
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            if (size == 0) {
+                tail = newNode;
+            }
+        } else if (index == size) {
+            newNode.prev = tail;
+            if (tail != null) {
+                tail.next = newNode;
+            }
+            tail = newNode;
+            if (size == 0) {
+                head = newNode;
+            }
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index - 1; i++) {
+                current = current.next;
+            }
+            newNode.next = current.next;
+            newNode.prev = current;
+            if (current.next != null) {
+                current.next.prev = newNode;
+            }
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValid(index);
+
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
```

Comment:
Don't iterate your nodes from the head to the tail or vice versa. You can iterate only the first or second half of the list, depending on the index value. This optimization is already applied, but you should encapsulate it into a separate `findNodeByIndex` method to avoid code duplication.